app.rb

```diff
@@ -175,6 +175,69 @@ def respond_with(format, category, responses)
     end
   end
 
+  def form_xml_for_settings(category, tag, responses)
+    builder = Nokogiri::XML::Builder.new do |xml|
+      xml.send(category) {
+        responses.each do |response|
+          case category
+          when :providers
+             xml.provider(tag => response)
+          when :servers
+             xml.server(tag => response)
+          when :images
+             xml.image(tag => response)
+          when :regions
+             xml.region(tag => response)
+          end
+        end
+      }
+    end
+    return builder
+  end
+
+  get '/v1/providers.?:ext?' do
+    validate_params_ext
+    
+    settings_xml = form_xml_for_settings(:providers, :name, settings.providers)
+    responses = settings_xml.doc.css("providers>provider")
+
+    respond_with params[:ext], :providers, responses 
+    
+  end
+ 
+  get '/v1/:provider/servers/types.?:ext?' do
+    validate_params_ext
+    validate_params_provider    
+    
+    settings_xml = form_xml_for_settings(:servers, :type, settings.server_types)
+
+    responses = settings_xml.doc.css("servers>server")
+
+    respond_with params[:ext], :servers, responses
+  end
+
+  get '/v1/:provider/images/states.?:ext?' do
+    validate_params_ext
+    validate_params_provider    
+    
```

Comment:
This should not really be a dynamic implementation. The life-cycle has 4 defined states: `active`, `inactive`, `deprecated`, and `deleted`. Whether or not we have `deleted` images in any given framework is immaterial. I agree that a user should have to implicitly know the states thus a query API for providing the defined states is good. But this should be framework independent, i.e. `/v1/images/states.?:ext?` and the should return the static list of the defined states, no data processing needed.

Reply:
Well, we do need to check that `:ext` is valid, but I agree it shouldn't be provider-specific.